vcx/libvcx/src/messages/get_message.rs

```diff
@@ -303,53 +310,48 @@ impl Message {
     }
 
     pub fn decrypt(&self, vk: &str) -> Message {
-        // TODO: must be Result
+        // TODO: change to Result
+
+        // Allocate a new object to hold the output of this function.
         let mut new_message = self.clone();
+        // If the payload isn't None...
         if let Some(ref payload) = self.payload {
-            let decrypted_payload = match payload {
-                MessagePayload::V1(payload) => {
-                    if let Ok(payload) = Payloads::decrypt_payload_v1(&vk, &payload) {
-                        Ok(Payloads::PayloadV1(payload))
-                    } else {
-                        warn!("fallback to Payloads::decrypt_payload_v12 in Message:decrypt for MessagePayload::V1");
-                        serde_json::from_slice::<serde_json::Value>(&to_u8(payload)[..])
-                            .map_err(|err| VcxError::from_msg(VcxErrorKind::InvalidMessagePack, format!("Cannot deserialize MessagePayload: {}", err)))
-                            .and_then(|json| Payloads::decrypt_payload_v12(&vk, &json))
-                            .map(|json| {
-                                (
-                                    json.type_,
-                                    match json.msg {
-                                        serde_json::Value::String(_str) => _str,
-                                        value => value.to_string()
-                                    }
-                                )
-                            })
-                            .map(|(type_, payload)|
-                                Payloads::PayloadV2(PayloadV2 {
-                                    type_,
-                                    id: ::utils::uuid::uuid(),
-                                    msg: payload,
-                                    thread: Default::default(),
-                                })
-                            )
-                    }
+            // Try to decrypt it in one of the standard ways.
+            if let Ok((data, _thread, v12)) = Payloads::decrypt_helper(vk, payload) {
+                // The next few lines are horribly unidiomatic Rust, but I can't figure out a cleaner
+                // way to write it. Need a Rustacean to polish...
+                if let Some(typ) = v12 {
+                    let x = Payloads::PayloadV2(PayloadV2 {
+                        type_: typ,
+                        id: ::utils::uuid::uuid(),
+                        msg: data,
+                        thread: Default::default(), // threads were not supported in v12
+                    });
+                    new_message.decrypted_payload = ::serde_json::to_string(&x).ok();
+                } else {
+                    new_message.decrypted_payload = ::serde_json::to_string(&data).ok();
```

Comment:
Is the `data` already a string?

Reply:
Yes, but the code I was refactoring had this same strange logic in it, and I kept it because I wondered if I was misunderstanding something. I went back and studied it, and I agree that it is a redundant conversion. Removing it now.